src/librustc/ty/item_path.rs

```diff
@@ -151,9 +151,23 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
                 }
             }
 
-            cur_path.push(self.def_key(cur_def)
-                              .disambiguated_data.data.get_opt_name().unwrap_or_else(||
-                Symbol::intern("<unnamed>").as_str()));
+            let mut cur_def_key = self.def_key(cur_def);
+
+            // For a UnitStruct or TupleStruct we want the name of its parent rather than <unnamed>.
+            if let DefPathData::StructCtor = cur_def_key.disambiguated_data.data {
+                let parent = DefId {
+                    krate: cur_def.krate,
+                    index: cur_def_key.parent.expect("DefPathData::StructCtor missing a parent"),
+                };
+
+                cur_def_key = self.def_key(parent);
+            }
+
+            let data = cur_def_key.disambiguated_data.data;
+            let symbol =
+                data.get_opt_name().unwrap_or_else(|| Symbol::intern("<unnamed>").as_str());
+            cur_path.push(symbol);
```

Comment:
I'm not sure about this.

Reply:
@eddyb what specifically are you unsure about? The use of `<unnamed>`? I wonder if/when this can ever arise. Perhaps some cases like this?

```rust
struct Foo;
impl Foo {
    #[deprecated]
    fn bar(&self); // referenced like `Foo::bar`
}
```

also maybe something like this

```rust
fn foo() {
    let x = || {
        #[deprecated]
        fn bar() { }
        bar();
    };
}
```

It'd definitely be good to include those two cases in the test harness, if nothing else.

Reply:
The first should already addressed by `MethodTester` in `deprecation-lint.rs` which includes testing all the UFCS forms. 

I've added a test for the latter, which prints `use of deprecated item 'this_crate::test_fn_closure_body::{{closure}}::bar` which seems okay?

Reply:
Ah, I see that the `<unnamed>` only applies at the tip (and that it was here anyway). I guess @eddyb was likely complaining more about the fact that it now skips to the parent of a `StructCtor`. That made me nervous too, except that it is part of `push_visible_item_path`, which seems like it's a "user-facing printout" function, and in that case this *does* strike me as the right behavior.

Reply:
Ah, but my premise is totally false. We invoke `try_push_visible_item_path` from the normal `item_path`.

Reply:
In that case, I might rather do this "skip to parent of struct ctor" in the caller.

Reply:
Though really this is more evidence that `item_path_str` is just serving too many clients right now. Let me go do some ripgrep around -- I may be remembering wrong -- but I feel like it's used in a lot of places, some of which may want precision and some of which do not.

Reply:
>  In that case, I might rather do this "skip to parent of struct ctor" in the caller.

@Ryman do you see what I mean by this? thoughts?

Reply:
@nikomatsakis Yeah, I understand, I was waiting for a followup to 'Let me go do some ripgrep around' incase you had other ideas :P I think adding it to the caller could make sense here as we already do so in other cases in the only caller that I can find [see here](https://github.com/rust-lang/rust/blob/master/src/librustc/ty/item_path.rs#L205).

Reply:
To be honest I found it hard to assess all the places that it's being used. Still, that case you showed seems to suggest that we're already doing a similar amount of suppression.